src/dune_file.ml

```diff
@@ -159,17 +162,35 @@ module Pps_and_flags = struct
     let decode =
       let+ l, flags =
         until_keyword "--"
-          ~before:(plain_string (fun ~loc s -> (loc, s)))
-          ~after:(repeat string)
+          ~before:String_with_vars.decode
+          ~after:(repeat String_with_vars.decode)
+      and+ syntax_version = Syntax.get_exn Stanza.syntax
       in
+      let check_allowed_variable flag =
+        if String_with_vars.has_vars flag then
+          Syntax.Error.since (String_with_vars.loc flag)
+            Stanza.syntax
+            (1, 10)
+              ~what:"Using variables in pps flags"
+      in
+      let flags = Option.value flags ~default:[] in
+      if syntax_version < (1, 10) then
+        List.iter ~f:check_allowed_variable flags;
       let pps, more_flags =
-        List.partition_map l ~f:(fun (loc, s) ->
-          if String.is_prefix s ~prefix:"-" then
+        List.partition_map l ~f:(fun s ->
+          if String_with_vars.is_prefix s ~prefix:"-" then begin
+            if syntax_version < (1, 10) then
+              check_allowed_variable s;
             Right s
-          else
-            Left (loc, Lib_name.of_string_exn ~loc:(Some loc) s))
+          end else
+             let loc = String_with_vars.loc s in
+             match String_with_vars.text_only s with
+             | None -> no_templates loc "in the ppx library names"
+             | Some txt ->
+               Left (loc, Lib_name.of_string_exn ~loc:(Some loc) txt)
+        )
       in
-      (pps, more_flags @ Option.value flags ~default:[])
```

Comment:
( Btw, the final flags are "swapped": when we write "-arg1 -arg2 -- arg3 arg4" in the dune file the preprocessor will read "arg3 arg4 -arg1 -arg2" ; was it on purpose ? )

Reply:
Hmm, no that does look like a bug.

Reply:
Sorry, I actually misread the order is right.